compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -201,6 +201,10 @@ pub struct ResolverAstLowering {
 
     /// Lints that were emitted by the resolver and early lints.
     pub lint_buffer: Steal<LintBuffer>,
+
+    /// Information about functions signatures for delegation items expansion
+    pub has_self: LocalDefIdSet,
+    pub fns_arguments_count: FxHashMap<DefId, usize>,
```

Comment:
Please, make recursive delegation (delegating to another delegation item) an error.
These tables are not filled for delegation items, so recursive delegation items won't use correct values here.

They can be made to work by making `has_self` and `fn_argument_count` queries so they could be called recursively, but it's probably a task for a different PR.